esp/logging/logginglib/modularlogagent.cpp

```diff
@@ -0,0 +1,1526 @@
+    }
+    return getFile(key);
+}
+
+/**
+ * Determines if a new physical file is required to process the update reuest. A new file is
+ * required if:
+ *   - no file I/O object exists in the file
+ *   - daily rollover is enabled and the last write occurred on a different date than now
+ *   - size rollover is enabled and the new content will cause the size threshold to be
+ *     exceeded; the first write to a file will not require a new file regardless of size
+ *
+ * The variable map may be updated with system variables related to new file naming:
+ *   - creationVarName will be added when a new file is required and the pattern references it;
+ *     the value added will support a custom timestamp, even when not configured to be custom
+ */
+bool CFileTarget::needNewFile(size_t contentLength, const File& file, const Pattern& pattern, Variables& systemVariables) const
+{
+    const char* creationValue = m_creationDateTimeFormat;
+    bool needed = !haveFile(file);
+    if (!file.m_io) // implies needed is true
+    {
+        creationValue = m_creationDateFormat;
+    }
+    if (!needed && DailyRollover == m_rolloverInterval)
+    {
+        unsigned ly, lm, ld, ty, tm, td;
+        CDateTime now;
+        now.setNow();
+        file.m_lastWrite.getDate(ly, lm, ld, false);
+        now.getDate(ty, tm, td, false);
+        needed = (ly != ty || lm != tm || ld != td);
+    }
+    if (!needed && m_rolloverSize)
+    {
+        size_t currentSize = file.m_io->size();
+        if (m_header.length() != currentSize)
+            needed = ((currentSize + contentLength) > m_rolloverSize);
+    }
+    if (needed && pattern.contains(creationVarName, nullptr))
+        systemVariables[creationVarName] = creationValue;
+    return needed;
+}
+
+bool CFileTarget::haveFile(const File& file) const
+{
+    bool result = false;
+    if (file.m_io)
+    {
+        try
+        {
+            Owned<IFile> tmp(createIFile(file.m_filePath));
+            result = tmp->exists();
+            if (!result)
+                iwarnlog(Major, "open file '%s' does not exist", file.m_filePath.str());
+        }
+        catch(IException* e)
+        {
+            StringBuffer msg;
+            ierrlog(Major, "exception checking existence of file '%s': %s", file.m_filePath.str(), e->errorMessage(msg).str());
+            e->Release();
+        }
+    }
+    return result;
+}
+
+/**
+ * Assuming a valid file path can be generated from the given pattern and variables, attempts
+ * to create and open a new file. If the generated path references an existing file, the path
+ * will be adjusted in an attempt to generate a new, unused file name:
+ *   - a requested custom creation timestamp that peformed a date subsitution will retry with
+ *     a date and time substitution
+ *   - all other requests will use createUniqueFile
+ */
+bool CFileTarget::createNewFile(File& file, const Pattern& pattern, Variables& systemVariables) const
+{
+    Owned<IFile> newFile;
```

Comment:
I think this instance of newFile is redundant.

Reply:
Me too.